File: freecodecamp/structy-graphs/1_depth_first.js
// Graph structure

// a
// ├── b
// │   └── d
// │       └── f
// └── c
//     └── e

// Edges:
// a -> b, c
// b -> d
// c -> e
// d -> f

const depthFirstPrint = (graph, source) => {
    const stack = [ source ];

    while (stack.length > 0){
        const current = stack.pop();
        console.log("current:", current)

        for (let neighbor of graph[current]){
            stack.push(neighbor);
        } 
    }

};

const graph = {
  a: ['b', 'c'],
  b: ['d'],
  c: ['e'],
  d: ['f'],
  e: [],
  f: []
};

depthFirstPrint(graph, 'a'); // a, c, e, b, d, f

File: freecodecamp/structy-graphs/1_depth_first.py
# Graph structure
#
# a
# ├── b
# │   └── d
# │       └── f
# └── c
#     └── e
#
# Edges:
# a -> b, c
# b -> d
# c -> e
# d -> f

def depth_first_print(graph, source):
    stack = []
    stack.append(source)

    while stack:
        current = stack.pop()
        print("current:", current)

        for n in graph[current]:
            stack.append(n)


graph = {
  "a": ["b", "c"],
  "b": ["d"],
  "c": ["e"],
  "d": ["f"],
  "e": [],
  "f": []
}

depth_first_print(graph, "a"); # a, c, e, b, d, f